doc/api/esm.md

```diff
@@ -268,8 +268,24 @@ import feature from 'es-module-package/features/x.js';
 If a package has no exports, setting `"exports": false` can be used instead of
 `"exports": {}` to indicate the package does not intend for submodules to be
 exposed.
-This is just a convention that works because `false`, just like `{}`, has no
-iterable own properties.
+
+Any invalid exports entries will be ignored. This includes exports not
```

Comment:
Should we perhaps warn on them, instead of silently ignoring them?

Reply:
The problem with warnings in Node.js is that when you warn for behaviour of packages in node_modules, there isn't much the user can do. This is in contrast to the browser, where warnings are for developers not users (console = developers), wheras in Node.js the users get the console.

So if we just spit out these warnings and we have a future-compatibility path, then it's just going to be spam, so I'd be against that.

Perhaps we can explicitly make these debug messages though under `NODE_DEBUG=exports` or similar in a follow-on PR.

Reply:
That’s a fair point, but it’s the same with that event emitter memory leak issue.

Reply:
It's not quite the same as the memory leak warning because we *expect* that there are invalid entries and ignoring them on older versions is exactly what should be happening on those. I would compare it to warning on unknown `package.json` properties. It would catch `mian: lib.js` but it would be much more likely to warn on all kinds of stuff that is perfectly correct.

Reply:
It’s the same in that it’s not actionable to the top level app, because it originates in an installed package.

The same is true for npm audit warnings.

I think this kind of warning is important for encouraging proper ecosystem behavior.

Reply:
The difference here is we are thinking about forwards-compatibility. On forwards compatibility paths where we add new features to "exports" that were previously validation failures, we don't want those validation failures then causing warning output in old Node.js versions that are running those packages.

That is, the warning is not something that should be corrected, because it is correct in modern Node.js, just incorrect in older versions.

I've been thinking of trying to ensure we provide more context on the validation error failures when they are matched, and may push up a commit along these lines shortly improving the error messages to know how to correct things.

If we want the "general validation" I would suggest that should be a special separate feature that runs under a top-level Node.js flag and validates all "exports" entries, where the user is then expecting the onslaught of messages and can deal with them.

Reply:
Why don’t we? If you’re using a node version that doesn’t match what one of your deps requires, you should get notified *somehow*.

Reply:
Because it's fine to ship modern features as long as they don't break older apps that don't support those features. The older app still works - the user doesn't need to be notified that a dependency needs changing, because the dependency is correct too. There is no action to be taken.

Reply:
That's not what we're saying. It's the equivalent of warning on the following code:

```js
if (fs.newAPI) { // "WARNING, fs doesn't include newAPI"
  fs.newAPI();
} else {
  fallbackBehavior();
}
```

You'd be warning on the feature check, not on the actual use. Equivalent of the above:

```js
{
  // we don't want to trigger warnings in versions that don't support future:api yet
  "/foo": ["future:api", "./fallback.mjs"]
}
```

Reply:
ah. That’s a valid export with a not-found module. I’m talking about an *invalid* export.

Reply:
The latest commit I just added gives the following for invalid export resolutions - 

eg -

```json
{
  "exports": {
    "./x": "in:valid"
  }
}
```

`import 'pkg/x'` will then provide the error:

```
Error: Cannot resolve package exports target 'in:valid' matched for './x' in C:\Users\Guy\Projects\node\node_modules\pkg\package.json, imported from C:\Users\Guy\Projects\node\x.mjs
    at Loader.resolve [as _resolve] (internal/modules/esm/default_resolve.js:59:13)
    at Loader.resolve (internal/modules/esm/loader.js:73:33)
    at Loader.getModuleJob (internal/modules/esm/loader.js:149:40)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:43:40)
    at link (internal/modules/esm/module_job.js:42:36) {
  code: 'ERR_MODULE_NOT_FOUND'
```

Reply:
what about for any processing of that package.json? I’d expect a warning when the package is read, not just when that path is imported.

Reply:
You could imagine a package that provides one export `pkg/newfeature` for modern Node.js, and another `pkg/legacyfeature` for legacy Node.js. If we validate all exports on load we get the same issue discussed above with warnings for things that aren't used.

As mentioned, I'd be open to a top-level flag for this kind of global validation, but I'd be against making it the default behaviour for the reason that it is simply entirely unactionable to the user - you can't even post a PR to the dependency package itself like other Node.js warnings for node_modules.

Reply:
again, I’m talking about invalid ones, not nonexistent ones.

I think it’s appropriate to always warn when using a package whose package.json has *any* exports that can’t be used in my node version. The actionable item is “stop using that package” or “upgrade node”.

Reply:
Along the lines of Jan's comment above, this is like saying that an API exporting:

```js
exports.legacy = function () {
  // uses fs.oldAPI();
}
exports.modern = function () {
  // uses fs.newAPI();
}
```

should warn in legacy environments that the `modern` function exists just because it might call an undefined core library if you run `pkg.modern()`.

There is nothing wrong with `pkg.modern` function existing if it isn't called. Similarly there should be nothing wrong with defining exports that only apply to modern use cases.

As I've mentioned, I'm happy to have a global validation mode as opt-in, but it shouldn't be the default for these reasons.